rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+
+---
+
+**Example**:
+
+A block expresses that it needs a URL hosting an image (presumably so that it can render it):
+
+```json
+{
+  "properties": {
+    "image_url": "string"
+  }
+}
+```
+
+The embedding application is then able to search across its data to find entities that satisfy this requirement, such as:
+
+```json
+{
+  "image_url": "https://dummyimage.com/300.jpg",
+  ... // some other fields
+}
+```
+
+It is then able to pass the value to the block, satisfying its requirements for input data.
+
+---
+
+This works most easily when the data uses the same identifier (key in the object) as the block expects, but problems arise if people have different ways of describing compatible data. For example, the embedding application might have an entity that looks like:
+
+```json
+{
+  "display_picture": "https://dummyimage.com/300.jpg",
+  ... // some other fields
+}
+```
+
+This is an example of a lack of **consensus**, where two parties don't agree on how to describe the data in question. At this point, there isn't a clear route to deciding compatibility. Block schemas could perhaps be modified so that the block could express it needs something that matches a [regex to ensure the URL has an image extension](https://stackoverflow.com/questions/42853961/regex-that-matches-image-links-but-exclude-normal-url), and the embedding application could ignore the key and just match on values, but this:
+
+- is potentially flakey ([here's an example](https://dummyimage.com/300) of a URL that points to an image but _does not_ have an extension)
+- ends up being very domain-specific
+- introduces a large maintenance burden
+- is generally not reusable, possibly leading to similar bespoke logic being defined in multiple places, with possible subtle but incompatible variations
+
+Thus the main reason for following RFC:
+
+> 💬 Removing some flexibility to establish a stronger contract between the block developer, embedding application, and end-user, allows for more powerful run-time inference and general confidence of all parties. Furthermore, creating a well-defined hierarchy of types, that are reusable and discoverable, encourages emergent standardization and, in turn, leads to higher-quality interoperable parts.
+
+In the following document we outline an approach to being more specific about the types that describe data. Through the following definitions of types, the community will be able to **gradually converge** upon a **consensus** of how to describe information.
+
+## Motivating implications
+
+**embedding applications:**
+
+- are required to do less guesswork about what the block is asking for
+- require less user intervention (or hard-coded logic/guesses) to map X to Y
+
+**blocks:**
```

Comment:
Firstly, I 100% agree with the issue you've identified, I think we should be a lot more up-front and clear about how the potential is mostly only realised if we can guide people into re-using types, etc.

I don't fully agree that there isn't value without that happening though, I think a lot of the bullet points are still true even if there isn't common overlap in the types between EAs and Blocks, because this RFC gives more ways for mismatched descriptions to be addressed. It is easier for an EA to map disjoint descriptions of "image" to one when they have things (types) to use, and more information (embedded in those types) available to make the decision.

Reply:
What more information is carried by a separate property definition compared to an in-line property definition? 

How does an EA map between two different property definitions without user intervention?

Reply:
> What more information is carried by a separate property definition compared to an in-line property definition?

Property Types require more information than we were previously asking for in inline things. For one, they will be associated with Data Types which later on _may_ encapsulate more semantic meaning, but let's disregard that as it's not part of this RFC. The main thing is that in this case Property Types are matched up with descriptions, and we can make it possible to see the type of contexts they're used in (by showing blocks that use them in the schemas for example). This type of contextualisation is not possible/enforced when your property is just a key. 

Using URIs as identifiers also lets us (even programmatically if such things are stored) disambiguate between properties that have the same name but different semantic meaning. 

Now a lot of it isn't really able to be programmatically inferred on first discovery, so I don't think some super general EA will look at the `description` of `image_url` and know that it's on about a profile picture or something. However for EAs that are including types in the source-code, or ones that are storing user decisions, this proposal provides new routes. i.e. if I've seen `image_url` before and the user has been able to encapsulate the semantic meaning in a way that I can store, when I see another block using the same property type I can reuse that stored meaning, as opposed to just seeing another string that's equivalent and having no idea if it's on about the same thing.

> How does an EA map between two different property definitions without user intervention?

I make no claims about the implementation details here, or that there won't be user intervention. I'm only suggesting that there are implementation possibilities that are now available that weren't before. If a _user stores_ some mapping of property definitions, then that can still be reused without user intervention if the property appears in some previously unseen block, etc.

Reply:
You have swayed me, thank you :D I am adding the suggestion. I think there are benefits without type sharing but I do not think the bullet points are the ones. As you pointed out all of my examples (regarding programmatic things) actually require reuse at some point!